### PR TITLE
Feature/pagination

### PR DESCRIPTION
**Pull request by j-tulforov**
- Added new component:
   - Pagination

- Updated some theme properties according to [layout](https://www.figma.com/file/66L3gxbUxVhqhXQjy4xzOz/Web?node-id=12355%3A60813) and designer command (instruction).

- Add new colors & props to text component

image

image

image

### CODE REVIEW

File: packages/button/src/ButtonStyles.tsx [hunk 1]
```diff
@@ -178,7 +178,7 @@ export const ButtonStyle = styled.button<InjectedProps>`
   position: relative;
   background: transparent;
   font-family: inherit;
-  font-weight: 800;
+  font-weight: 400;
```
Comment (Drujininii) [inline packages/button/src/ButtonStyles.tsx:181]: поясни пожалуйста как обратную совместимость сохраняет. разве не поменяются все кнопки?

File: packages/pagination/src/Pagination.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    [pagesCount, currentPage, siblingCount]
+  )
+
+  if (pagesCount <= 0) {
+    return null
+  }
+
+  const onPageItemClick = (page: number, e: React.MouseEvent) => {
+    onItemClick(page, e)
+    setCurrPage(page)
```
Comment (Drujininii) [inline packages/pagination/src/Pagination.tsx:44]: без точек с запятой конечно больно)

File: packages/pagination/src/Pagination.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      <PaginationItem
+        disabled={currentPage === 1}
+        icon={<ArrowLeft />}
+        onClick={onPreviousClick}
+      />
+
+      {showingPages.map((page, index) => {
+        const isDisabled = page === '...'
+        const variant: PaginationItemProps['variant'] =
+          page === currentPage ? 'active' : 'default'
```
Comment (Drujininii) [inline packages/pagination/src/Pagination.tsx:74]: можем в enum вынести?

File: packages/pagination/src/PaginationItem.tsx [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import React from 'react'
+import { PaginationItemStyle } from './PaginationItemStyles'
+import { PaginationItemProps } from './types'
+
+const PaginationItem: React.FC<PaginationItemProps> = (props) => {
+  const { icon, variant = 'default', ...rest } = props
```
Comment (Drujininii) [inline packages/pagination/src/PaginationItem.tsx:6]: и сюда enum

File: packages/pagination/src/getShowingPages.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+export const range = (start: number, end: number) => {
+  const length = end - start + 1
+  return Array.from({ length }, (_, i) => start + i)
+}
+
+const getShowingPages = (
+  pagesCount: number,
+  currPage = 1,
+  siblingCount: SiblingsCount = 1
```
Comment (Drujininii) [inline packages/pagination/src/getShowingPages.ts:26]: почему именно siblingCount? почему не пропс с количеством элементов в пагинаторе?

File: packages/pagination/src/Pagination.tsx (hunk 2 above)
Comment (j-tulforov) [reply packages/pagination/src/Pagination.tsx:44]: мда, но скоро это починять единым линтером

File: packages/pagination/src/Pagination.tsx (hunk 3 above)
Comment (j-tulforov) [reply packages/pagination/src/Pagination.tsx:74]: Хорошо что подметил, уже есть enum, просто нужно его тут использовать

File: packages/button/src/ButtonStyles.tsx (hunk 1 above)
Comment (alx-khramov) [reply packages/button/src/ButtonStyles.tsx:181]: @Drujininii we should use english for reviews, PRs, etc. Especially because this is a public repo.

Comment (j-tulforov) [reply packages/button/src/ButtonStyles.tsx:181]: Discussed it with Stanislav,`font-weight: 400` should be another `variant` of `Button`, thanks for point, I was misunderstanding.

Anyway, he said `font-weight` for buttons should be **700**, not **800**. It's ok that there would not be backward compatibility for this.

File: packages/button/src/ButtonStyles.tsx [hunk 6]
```diff
@@ -178,7 +178,7 @@ export const ButtonStyle = styled.button<InjectedProps>`
   position: relative;
   background: transparent;
   font-family: inherit;
-  font-weight: 800;
+  font-weight: 700;
```
Comment (j-tulforov) [inline packages/button/src/ButtonStyles.tsx:181]: Discussed with Stanislav, he said font-weight for buttons should be 700, not 800. It's ok that there would not be backward compatibility for this. [Proofs](https://www.figma.com/file/0j4EJ3Ra7OPBeEKLXFQiGM/LIDO-Design-System?node-id=10497%3A2582)

File: packages/button/src/ButtonStyles.tsx [hunk 7]
```diff
@@ -31,14 +31,14 @@ const sizes = {
...
-    font-size: ${({ theme }) => theme.fontSizesMap.xxs}px;
+    font-size: ${({ theme }) => theme.fontSizesMap.xs}px;
     border-radius: ${({ theme }) => theme.borderRadiusesMap.sm}px;
     padding: ${({ $square }) => ($square ? '16px' : '16px 24px')};
     min-width: ${({ $square }) => ($square ? '0' : '100px')};
   `,
   md: css<InjectedProps>`
     line-height: 1em;
-    font-size: ${({ theme }) => theme.fontSizesMap.xs}px;
+    font-size: ${({ theme }) => theme.fontSizesMap.sm}px;
```
Comment (j-tulforov) [inline packages/button/src/ButtonStyles.tsx:41]: Discussed with Stanislav, found out that sizes for these buttons are incorrect. [Proofs](https://www.figma.com/file/0j4EJ3Ra7OPBeEKLXFQiGM/LIDO-Design-System?node-id=10497%3A2582)

File: packages/pagination/src/getShowingPages.ts (hunk 5 above)
Comment (j-tulforov) [reply packages/pagination/src/getShowingPages.ts:26]: Firstly I also thought about pagesCount, but after research I figured out that https://mui.com/material-ui/react-pagination/#pagination-ranges use `siblingCount` & `boundaryCount`. That's why for better developer experience, I decided to choose `siblingCount`

File: packages/pagination/src/Pagination.tsx (hunk 3 above)
Comment (Drujininii) [reply packages/pagination/src/Pagination.tsx]: could you mark as todo?

**Review by alx-khramov — COMMENTED** [review]
Checked the UI in Storybook. Looks and feels beautiful.
But it is not quite clear what is SiblingCount and why it can only be 0 or 1.

File: packages/pagination/src/getShowingPages.test.ts [hunk 8]
```diff
@@ -0,0 +1,221 @@
+import getShowingPages, {
+  range,
+  numberSort,
+  arrayPusher,
+} from './getShowingPages'
+
+describe('getShowingPages при siblingCount=1', () => {
```
Comment (mymphe) [inline packages/pagination/src/getShowingPages.test.ts:7]: Please use english

File: packages/pagination/src/getShowingPages.test.ts [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+      test(`pagesCount = ${value}`, () => {
+        const received = getShowingPages(value, 1, 1)
+
+        expect(received).toStrictEqual(expected)
+      })
+    }
+  })
+})
+
+describe('getShowingPages при siblingCount=0', () => {
```
Comment (mymphe) [inline packages/pagination/src/getShowingPages.test.ts:88]: Please use english

File: packages/pagination/src/getShowingPages.test.ts [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+  test('start=1, end=5', () => {
+    const received = range(1, 5)
+    const expected = [1, 2, 3, 4, 5]
+
+    expect(received).toStrictEqual(expected)
+  })
+})
+
+describe('arrayPusher', () => {
+  test('Добавление 5 элементов', () => {
```
Comment (mymphe) [inline packages/pagination/src/getShowingPages.test.ts:179]: Please use english

File: packages/pagination/src/getShowingPages.test.ts [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+    pusher(pages, 10)
+    pusher(pages, 8)
+    const count = pusher(pages, 5)
+
+    expect(count).toBe(5)
+    expect(count).toEqual(pages.length)
+    expect(pages).toEqual([1, 2, 10, 8, 5])
+  })
+
+  test('Добавление несколько ...', () => {
```
Comment (mymphe) [inline packages/pagination/src/getShowingPages.test.ts:196]: Please use english

File: packages/pagination/src/getShowingPages.test.ts [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+    const count = pusher(pages, 1)
+
+    expect(count).toBe(3)
+    expect(count).toEqual(pages.length)
+    expect(pages).toEqual(['...', 1, '...'])
+  })
+})
+
+describe('numberSort', () => {
+  test('Сортировка номеров', () => {
```
Comment (mymphe) [inline packages/pagination/src/getShowingPages.test.ts:213]: Please use english

File: packages/pagination/src/getShowingPages.test.ts (hunk 8 above)
Comment (j-tulforov) [reply packages/pagination/src/getShowingPages.test.ts:7]: ops)

File: packages/pagination/src/Pagination.tsx (hunk 3 above)
Comment (j-tulforov) [reply packages/pagination/src/Pagination.tsx]: I have already do it.

**Comment by j-tulforov** [comment]
> Checked the UI in Storybook. Looks and feels beautiful. But it is not quite clear what is SiblingCount and why it can only be 0 or 1.

@alx-khramov https://github.com/lidofinance/ui/pull/314#discussion_r851255805

By default, at least in my UX ([also DX, developer experience](https://mui.com/material-ui/react-pagination/)), we show these elements in pagination:
- current element
- current's siblings (**curr-1**, **curr+1**)
- first element
- last element
- two ellipsis (...)

For example:
**5 => 1, ..., 4, 5, 6, …, 10**

So, at least we need 7 elements to show (+2, next & previous buttons). But in [layout](https://www.figma.com/file/66L3gxbUxVhqhXQjy4xzOz/Web?node-id=12355%3A61235), in mobile version we can show only 5. That's why we need to support also [variant without siblings](https://mui.com/material-ui/react-pagination/#pagination-ranges). In above example it will be **5 => 1, ..., 5, …, 10**.

Of course we can support more siblingsCount, 2, 3,etc. Right now, there is no need.

**Comment by alx-khramov** [comment]
@j-tulforov https://github.com/lidofinance/ui/pull/314#issuecomment-1101240254
Got it, thank you. It would be cool to describe it in the package README at least, or in the Storybook.
I saw 2 stories you made, but there is no detailed description of SiblingCount.

**Review by kozlovzxc — APPROVED** [review]
yarn.lock updates locks safe 👍

**Comment by j-tulforov** [comment]
> @j-tulforov [#314 (comment)](https://github.com/lidofinance/ui/pull/314#issuecomment-1101240254) Got it, thank you. It would be cool to describe it in the package README at least, or in the Storybook. I saw 2 stories you made, but there is no detailed description of SiblingCount.

@alx-khramov good point! Thanks for thats, truly, explanation was missing. Added it, plz check it, is it ok?

File: packages/pagination/src/getShowingPages.test.ts (hunk 12 above)
Comment (j-tulforov) [reply packages/pagination/src/getShowingPages.test.ts]: thanks for observation, done

File: packages/pagination/src/getShowingPages.test.ts (hunk 11 above)
Comment (j-tulforov) [reply packages/pagination/src/getShowingPages.test.ts]: done

File: packages/pagination/src/getShowingPages.test.ts (hunk 10 above)
Comment (j-tulforov) [reply packages/pagination/src/getShowingPages.test.ts]: done

File: packages/pagination/src/getShowingPages.test.ts (hunk 9 above)
Comment (j-tulforov) [reply packages/pagination/src/getShowingPages.test.ts]: done

**Review by mymphe — APPROVED** [review]

File: packages/pagination/src/Pagination.tsx [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+      {showingPages.map((page, index) => {
+        const isDisabled = page === '...'
+        const variant: PaginationItemProps['variant'] =
+          page === currentPage
+            ? PaginationItemVariant.active
+            : PaginationItemVariant.default
+
+        return (
+          <PaginationItem
+            key={`${page}${index}`}
```
Comment (Jabher) [inline packages/pagination/src/Pagination.tsx:84]: better to use something like 
```typescript
page === '...' ? `${page}${index}` : page
```

**Review by alx-khramov — APPROVED** [review]
Checked the `SiblingCount` and other props description in the readme. Looks good for me. Thank you for the work. 👍🏻

File: packages/pagination/src/Pagination.tsx (hunk 13 above)
Comment (j-tulforov) [reply packages/pagination/src/Pagination.tsx:84]: I guess it's more reliable? but didn't understand why.

File: packages/theme/src/base/colors.ts [hunk 14]
```diff
@@ -20,7 +20,7 @@ const colors = {
   warningHover: '#f0a431',
   warningContrast: '#fff',
 
-  success: '#61b75f',
+  success: '#53BA95',
```
Comment (Jabher) [inline packages/theme/src/base/colors.ts:23]: did you research of affected scope?

File: packages/text/src/TextStyles.tsx [hunk 15]
```diff
@@ -30,23 +30,52 @@ export const sizes = {
...
-  font-weight: 500;
-  ${(props) => sizes[props.$size]}
+  ${({ strong, italic, size }) => css`
+    font-style: ${italic ? 'italic' : 'normal'};
+    font-weight: ${strong ? 700 : 400};
+    margin: 0;
+    padding: 0;
+    color: ${getTextColor};
+    text-decoration: ${getTextDecoration};
+    ${sizes[size]}
```
[the remark this answers is not in the document]
Comment (j-tulforov) [reply packages/text/src/TextStyles.tsx:79]: only type-checking, If u think that need also "if", then it should be in all components, not only here

[the remark this answers is not in the document]
Comment (Jabher) [reply packages/text/src/TextStyles.tsx:79]: Thanks for the point. Will probably come with that later (maybe significantly later)

File: packages/pagination/src/Pagination.tsx (hunk 13 above)
Comment (Jabher) [reply packages/pagination/src/Pagination.tsx:84]: when you are replacing e.g. `1,2,3,...,9,10,11` with `1,2,3,...,10,11,12` (you can imagine different cases here, this is simplest one), if you are using `value & index` key you are getting every element (except first ones) re-rendered. However, if using `value` key you will get re-renders only for '...' elements

Comment (Jabher) [reply packages/pagination/src/Pagination.tsx:84]: however, I cannot say "key" is more reliable as it is only hinting for react to re-use already existing components for performance reasons - you can change all the properties, mis-hit the cache, and you will only get slightly slower renderer.

Anyway - it is just a comment, not insisting on enforcing it: we're not hunting on nanoseconds on pagination :)

Comment (j-tulforov) [reply packages/pagination/src/Pagination.tsx:84]: Ah yeah, doesn't think about situation when index is changing, u'r right.

Comment (j-tulforov) [reply packages/pagination/src/Pagination.tsx]: done

**Comment by j-tulforov** [comment]
@Jabher @alx-khramov if there is no any else comments, please, let's merge and publish it. I need this changes in my task.

**Comment by Jabher** [comment]
No more comments from me

**Pull request merged**